text/0000-public-private-dependencies.md

```diff
@@ -0,0 +1,199 @@
+possible to force this type to be distinct.  This means that users of the crate might
+accidentally start depending on that type to be compatible if the user starts to depend
+on the crate that actually implements that type.  The limitations from the previous
+answer apply (eg: you can currently overrule the restrictions).
+
+**Q: How does semver and depenencies interact?**<br>
+A: It is already the case that changing your own dependencies would require a semver
+bumb for your own library because your API contract to the outside world changes.  This
+RFC however makes it possible to only have this requirement for public dependencies and
+would permit cargo to prevent new crate releases with semver violations.
+
+# Detailed design
+[design]: #detailed-design
+
+There are a few areas that require to be changed for this RFC:
+
+* The compiler needs to be extended to understand when crate dependencies are
+  considered a public dependency
+* The `Cargo.toml` manifest needs to be extended to support declaring public
+  dependencies
+* The cargo publish process needs to be changed to warn (or prevent) the publishing
+  of crates that have undeclared public dependencies
+* crates.io should show public dependencies more prominently than private ones.
+
+## Compiler Changes
+
+The main change to the compiler will be to accept a new parameter that cargo
+supplies which is a list of public dependencies.  The compiler then emits
+warnings if it encounters private dependencies leaking to the public API of a
+crate.  `cargo publish` might change this warning into an error in its lint
+step.
+
+Additionally later on the warning can turn into a hard error in general.
+
+In some situations it can be necessary to allow private dependencies to become
+part of the public API.  In that case one can permit this with
+`#[allow(external_private_dependency)]`.  This is particularly useful when
+paired with `#[doc(hidden)]` and other already existing hacks.
+
+This most likely will also be necessary for the more complex relationship of
+`libcore` and `libstd` in Rust itself.
+
+## Changes to `Cargo.toml`
+
+The `Cargo.toml` file will be amended to support the new `public` parameter on
+dependencies.  Old cargo versions will emit a warning when this key is encountered
+but otherwise continue.  Since the default for a dependency to be private only
+public ones will need to be tagged which should be the minority.
+
+Example dependency:
+
+```toml
+[dependencies]
+url = { version = "1.4.0", public = true }
+```
+
+## Changes to Cargo Publishing
+
+When a new crate version is published Cargo will warn about types and traits that
+the compiler determined to be public but did not come from a public dependency.  For
+now it should be possible to publish anyways but in some period in the future it will
+be necessary to explicitly mark all public dependencies as such or explicitly
+mark them with `#[allow(external_private_dependency)]`.
```

Comment:
I for some reason thought you meant that `#[allow(external_private_dependency)]` needs to be reflected in the index.